## Nama  : Salfa Putri Lestari
## NIM   : 2404220002

Import Library

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.preprocessing import StandardScaler

Load Data

In [2]:
train = pd.read_csv('/content/data_training.csv')
test = pd.read_csv('/content/data_testing.csv')

print("Data training:")
print(train.head())

print("\nData testing:")
print(test.head())


Data training:
   fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.3             0.740         0.08             1.7      0.094   
1            8.1             0.575         0.22             2.1      0.077   
2           10.1             0.430         0.40             2.6      0.092   
3           12.9             0.500         0.55             2.8      0.072   
4            8.4             0.360         0.32             2.2      0.081   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 10.0                  45.0  0.99576  3.24       0.50   
1                 12.0                  65.0  0.99670  3.29       0.51   
2                 13.0                  52.0  0.99834  3.22       0.64   
3                  7.0                  24.0  1.00012  3.09       0.68   
4                 32.0                  79.0  0.99640  3.30       0.72   

   alcohol  quality    Id  
0      9.8        5  1366  
1      9.2     

pada tahap ini, dataset data_training dan data_testing dibaca dengan menggunakan fungsi pd.read_csv(). dan memunculkan dengan menampilkan beberapa baris pertama dataset dengan fungsi head(). hal ini bertujuan untuk memastikan data berhasil dimuat dengan benar serta membantu melihat struktur awal data seperti nama kolom, dan isi dataset.

Cek Data

In [3]:
print("\nInfo data training:")
print(train.info())

print("\nCek missing value:")
print(train.isnull().sum())


Info data training:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857 entries, 0 to 856
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         857 non-null    float64
 1   volatile acidity      857 non-null    float64
 2   citric acid           857 non-null    float64
 3   residual sugar        857 non-null    float64
 4   chlorides             857 non-null    float64
 5   free sulfur dioxide   857 non-null    float64
 6   total sulfur dioxide  857 non-null    float64
 7   density               857 non-null    float64
 8   pH                    857 non-null    float64
 9   sulphates             857 non-null    float64
 10  alcohol               857 non-null    float64
 11  quality               857 non-null    int64  
 12  Id                    857 non-null    int64  
dtypes: float64(11), int64(2)
memory usage: 87.2 KB
None

Cek missing value:
fixed acidity           0
volati

Setelah dataset berhasil ditampilkan dengan menggunakan fungsi info() untuk melihat jumlah data, tipe data setiap variabel, serta memastikan seluruh kolom sudah sesuai dengan kebutuhan analisis. Dan sementara itu fungsi isnul().sum() berfungsi untuk mengecek apakah terdapat missing value.

Menangani Missing Value

In [4]:
train = train.fillna(train.mean(numeric_only=True))


Berikut merupakan penanganan missing value menggunakan metode rata-rata. data kosong diisi menggunakan nilai rata-rata dari masing-masing kolom numerik agar jumlah data tetap terjaga.

Pisah Fitur dan Target

In [5]:
X = train.drop(['quality', 'Id'], axis=1)
y = train['quality']

Data kemudia dipisahkan menjadi variabel fitur (X) dan variabel target (y). Variabel fitur berisi karakteristik kimia wine seperti kadar alkohol, pH, density, sulphates, chlorides, dan variabel lainnya yang digunakan sebagai dasar prediksi. Sementara itu, variabel target berupa quality yang menjadi nilai yang ingin diprediksi oleh model. Kolom Id juga dihapus karena hanya berfungsi sebagai identitas data dan tidak memiliki hubungan langsung terhadap kualitas wine.

Split Data

In [6]:
X_train, X_val, y_train, y_val = train_test_split(
    X, y, test_size=0.2, random_state=42
)

tahap ini berfungsi untuk memastikan bahwa model tidah hanya menghafal data training, tetapi juga dapat mampu untuk melakukan prediksi dengan baik pada data baru.

Scaling

In [7]:
scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)


Tahap selanjutnya yaitu melakukan scaling dengan menggunakan fungsi StandarScaler untuk menyamakan rentang nilai antar variabel. Tahap ini penting karena beberapa fitur memiliki skala yang berbeda cukup jauh. dengan scaling hasilnya akan menjadi lebih seimbang sehingga model dapat bekerja lebih stabil dan optimal.

Buat Model

In [8]:
model = RandomForestClassifier(
    n_estimators=200,      # jumlah pohon
    max_depth=None,
    random_state=42
)

# Latih model
model.fit(X_train, y_train)

RandomForestClassifier(n_estimators=200, random_state=42)

Model pertama yang digunakan adalah Random Forest Classifier, yaitu metode yang menggabungkan banyak decision tree dan menentukan hasil berdasarkan voting terbanyak. Cara ini membuat model lebih stabil, mampu menangkap pola kompleks, dan mengurangi overfitting, sehingga cocok untuk klasifikasi seperti kualitas wine. Setelah dilatih, model diuji menggunakan data validasi. Accuracy digunakan untuk melihat ketepatan prediksi, confusion matrix untuk mengetahui kesalahan pada tiap kelas, dan classification report (precision, recall, f1-score) untuk evaluasi yang lebih detail. Jika nilainya tinggi, berarti model sudah cukup baik dalam mengenali kualitas wine.


Evaluasi Model

In [9]:
y_pred = model.predict(X_val)

print("\nAkurasi model:", accuracy_score(y_val, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_val, y_pred))
print("\nClassification Report:\n", classification_report(y_val, y_pred))


Akurasi model: 0.6162790697674418

Confusion Matrix:
 [[ 0  1  2  0  0]
 [ 0 45 22  0  0]
 [ 0 22 53  3  0]
 [ 0  2 11  8  0]
 [ 0  0  0  3  0]]

Classification Report:
               precision    recall  f1-score   support

           4       0.00      0.00      0.00         3
           5       0.64      0.67      0.66        67
           6       0.60      0.68      0.64        78
           7       0.57      0.38      0.46        21
           8       0.00      0.00      0.00         3

    accuracy                           0.62       172
   macro avg       0.36      0.35      0.35       172
weighted avg       0.59      0.62      0.60       172



/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Selanjutnya dilakukan evalusai menggunakan data validasi. fungsi predict() digunakan untuk menghasilkan prediksi kualitas wine, kemudia hasil prediksi dilakukan pembandingkan dengan data asli menggunakan accuracy_score. Nilai akurasi menunjukkan seberapa besar kemampuan model dalam memprediksi data dengan benar. Selain itu, confusion matrix digunakan untuk melihat jumlah prediksi benar dan salah pada setiap kelas kualitas wine. Classification report memberikan informasi tambahan berupa precision, recall, dan f1-score yang membantu mengevaluasi performa model secara lebih detail. Jika nilai evaluasi cukup tinggi, maka model dapat dikatakan berhasil mengenali pola kualitas wine dengan baik.

Prediksi Data Testing

In [10]:
# ambil id
test_id = test['Id']

# buang kolom id dari fitur
X_test = test.drop('Id', axis=1)

# scaling test
X_test = scaler.transform(X_test)

# prediksi
test_pred = model.predict(X_test)

Setelah model dinilai cukup baik, model digunakan untuk memprediksi dataset testing yang belum memiliki label kualitas. Kolom Id terlebih dahulu dipisahkan karena nantinya akan digunakan pada file output. Selanjutnya, fitur pada data testing discaling menggunakan scaler yang sama dengan data training agar konsisten. Setelah itu, model melakukan prediksi terhadap kualitas wine pada dataset testing

Hasil Prediksi

In [26]:
hasil_rf = pd.DataFrame({
    'Id': test_id,
    'quality': test_pred
})

# tampilkan hasil prediksi
print("\nHasil Prediksi Random Forest:")
print(hasil_rf.head())


Hasil Prediksi Random Forest:
     Id  quality
0   222        5
1  1514        5
2   417        5
3   754        5
4   516        5


Hasil prediksi menggunakan Random Forest menunjukkan bahwa sebagian besar data wine pada dataset testing diprediksi memiliki kualitas 5. Hal ini menandakan bahwa model mengenali pola karakteristik kimia wine pada data testing sebagai wine dengan kualitas menengah. Prediksi yang cenderung sama menunjukkan bahwa Random Forest cukup stabil dalam melakukan klasifikasi data.

REGRESI LOGISTIK

In [11]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

Membuat model regresi logistik

In [12]:
log_model = LogisticRegression(
    max_iter=5000,   # supaya iterasi cukup dan tidak error
    random_state=42
)

Selain Random Forest, digunakan juga metode Regresi Logistik untuk membandingkan performa model. Regresi Logistik merupakan metode klasifikasi yang bekerja berdasarkan probabilitas dan hubungan linear antar variabel. Model ini lebih sederhana dibanding Random Forest sehingga lebih mudah diinterpretasikan.

In [13]:
log_model.fit(X_train, y_train)

LogisticRegression(max_iter=5000, random_state=42)

Prediksi Data

In [14]:
log_pred = log_model.predict(X_val)

Evaluasi Model

In [15]:
print("=== HASIL REGRESI LOGISTIK ===")

print("\nAkurasi:")
print(accuracy_score(y_val, log_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_val, log_pred))

print("\nClassification Report:")
print(classification_report(y_val, log_pred))


=== HASIL REGRESI LOGISTIK ===

Akurasi:
0.5988372093023255

Confusion Matrix:
[[ 0  3  0  0  0]
 [ 0 55 12  0  0]
 [ 0 29 42  6  1]
 [ 0  1 14  6  0]
 [ 0  0  0  3  0]]

Classification Report:
              precision    recall  f1-score   support

           4       0.00      0.00      0.00         3
           5       0.62      0.82      0.71        67
           6       0.62      0.54      0.58        78
           7       0.40      0.29      0.33        21
           8       0.00      0.00      0.00         3

    accuracy                           0.60       172
   macro avg       0.33      0.33      0.32       172
weighted avg       0.57      0.60      0.58       172



/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


hasil akurasi menunjukkan sebesar 0.59, yang dimana nilainya lebih rendah dibandingkan model random forest sebelumnya.

Prediksi Data Testing

In [16]:
log_test_pred = log_model.predict(X_test)

Hasil Prediksi

In [25]:
hasil_logistik = pd.DataFrame({
    'Id': test_id,
    'quality': log_test_pred
})

print("\nHasil Prediksi Regresi Logistik:")
print(hasil_logistik.head())



Hasil Prediksi Regresi Logistik:
     Id  quality
0   222        5
1  1514        5
2   417        5
3   754        5
4   516        6


Hasil prediksi Regresi Logistik juga menunjukkan sebagian besar wine diprediksi memiliki kualitas 5, namun terdapat satu data yang diprediksi memiliki kualitas 6. Hal ini menunjukkan bahwa model melihat adanya sedikit perbedaan pola pada karakteristik wine tertentu sehingga menghasilkan prediksi yang berbeda.

**MEMBANDINGKAN HASIL RANDOM FOREST DAN REGRESI LOGISTIK**

In [18]:
from sklearn.metrics import accuracy_score

Hitung Akurasi Random Forest

In [19]:
rf_accuracy = accuracy_score(y_val, y_pred)

Hitung Akurasi Regresi Logistik

In [20]:
log_accuracy = accuracy_score(y_val, log_pred)

Hasil Akurasi

In [21]:
print("=== PERBANDINGAN MODEL ===")

print("\nAkurasi Random Forest : ", rf_accuracy)
print("Akurasi Regresi Logistik : ", log_accuracy)

=== PERBANDINGAN MODEL ===

Akurasi Random Forest :  0.6162790697674418
Akurasi Regresi Logistik :  0.5988372093023255


**Random Forest**
Hasil akurasi dari model Random Forest diperoleh sebesar 0.6163 atau sekitar 61.63%. hal ini menunjukan bahwa nilai tersebut dapat memprediksi kualitas wine dengan benar sekitar 61% dari seluruh data validasi. asil tersebut menunjukkan bahwa model sudah cukup baik dalam mengenali pola hubungan antara karakteristik kimia wine dengan kualitas wine.

**Regresi Logistik**
Selanjutnya untuk hasil akurasi model regresi logistik diperoleh hasil sebesar 0.5988 atau sekitar 59.88%. Hal ini menunjukkan bahwa model Regresi Logistik mampu memprediksi kualitas wine dengan benar sekitar 59% dari data validasi. Walaupun hasilnya cukup baik, performa model ini masih sedikit lebih rendah dibanding Random Forest.

Menentukan Model Terbaik

In [22]:
if rf_accuracy > log_accuracy:
    print("\nModel terbaik adalah RANDOM FOREST")

elif log_accuracy > rf_accuracy:
    print("\nModel terbaik adalah REGRESI LOGISTIK")

else:
    print("\nKedua model memiliki akurasi yang sama")


Model terbaik adalah RANDOM FOREST


Hasil perbandingan menunjukkan bahwa Random Forest memiliki akurasi lebih tinggi dibanding Regresi Logistik. Random Forest memperoleh akurasi sebesar 61,63%, sedangkan Regresi Logistik memperoleh 59,88%. Selisih ini menunjukkan bahwa Random Forest lebih baik dalam mengenali pola data kualitas wine pada dataset yang digunakan. Hal ini disebabkan Random Forest mampu menangani pola data yang lebih kompleks dan hubungan non-linear antar variabel, sehingga hasil prediksinya lebih stabil dan lebih akurat. Sementara itu, Regresi Logistik bekerja berdasarkan hubungan linear sehingga performanya lebih terbatas pada data dengan pola yang kompleks seperti dataset wine quality.

Memilih Model Terbaik

In [27]:
if rf_accuracy > log_accuracy:

    best_model = model
    best_prediction = test_pred
    best_model_name = "Random Forest"

else:

    best_model = log_model
    best_prediction = log_test_pred
    best_model_name = "Regresi Logistik"

Simpan Hasil Model Terbaik

In [28]:
hasil_terbaik = pd.DataFrame({
    'Id': test_id,
    'quality': best_prediction
})

In [29]:
hasil_terbaik.to_csv('hasilprediksi_terbaik_002.csv', index=False)

print(f"\nHasil prediksi dari model terbaik ({best_model_name}) berhasil disimpan!")


Hasil prediksi dari model terbaik (Random Forest) berhasil disimpan!
